jena-fuseki2/jena-fuseki-core/src/main/java/org/apache/jena/fuseki/server/DataService.java

```diff
@@ -64,22 +63,39 @@
     private final AtomicBoolean offlineInProgress       = new AtomicBoolean(false);
     private final AtomicBoolean acceptingRequests       = new AtomicBoolean(true);
 
-    /** Create a {@code DataService} for the given dataset. */
-    public DataService(DatasetGraph dataset) {
+    /** Builder for a new DataService. */
+    public static Builder newBuilder() { return new Builder(); }
+
+    /** Builder for a new DataService, with a dataset. */
+    public static Builder newBuilder(DatasetGraph dsg) {
+        return newBuilder().dataset(dsg);
+    }
+
+    /** Return a new builder, populated by an existing DatasetService */
+    public static Builder newBuilder(DataService dSrv) {
+        return new Builder(dSrv.dataset, dSrv.endpoints, dSrv.operationsMap, dSrv.authPolicy);
+    }
+
+    /** Create a {@code DataService} for the given dataset.
+     * @param authPolicy */
+    private DataService(DatasetGraph dataset, Map<String, EndpointSet> endpoints, ListMultimap<Operation, Endpoint> operationsMap, AuthPolicy authPolicy) {
         this.dataset = dataset;
+        this.endpoints = new HashMap<>(endpoints);
+        this.operationsMap = ArrayListMultimap.create(operationsMap);
+        this.authPolicy = authPolicy;
         counters.add(CounterName.Requests);
         counters.add(CounterName.RequestsGood);
         counters.add(CounterName.RequestsBad);
         // Need to call goActive(). Not automatic.
         //goActive();
     }
 
-    /** Clone this {@code DataService} except use a different dataset graph */
-    public DataService alter(DatasetGraph dataset) {
-        DataService dSrv2 = new DataService(dataset);
-        this.forEachEndpoint(ep->dSrv2.addEndpoint$(ep));
-        return dSrv2;
-    }
+//    /** Clone this {@code DataService} except use a different dataset graph */
```

Comment:
The commented out code might be useful later? If not we could remove it perhaps?